FTBFS on x32 architecture with SSE3 support enabled

Hello,

On debian x32 architecture, gdal FTBFS with the following error:

```
[ 95%] Linking CXX shared library libgdal.so
/usr/bin/cmake -E cmake_link_script CMakeFiles/GDAL.dir/link.txt --verbose=1
/usr/bin/c++ -fPIC -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -specs=/usr/share/dpkg/pie-compile.specs -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fvisibility=hidden -specs=/usr/share/dpkg/pie-link.specs -Wl,-z,relro -Wl,-z,now -Wl,--no-undefined -shared -Wl,-soname,libgdal.so.31 -o libgdal.so.31.0.0 @CMakeFiles/GDAL.dir/objects1.rsp  -Wl,-rpath,/usr/lib/x86_64-linux-gnux32/hdf5/serial: -lm -ldl /usr/lib/x86_64-linux-gnux32/libz.so /usr/lib/x86_64-linux-gnux32/libcurl.so /usr/lib/x86_64-linux-gnux32/libodbc.so /usr/lib/x86_64-linux-gnux32/libodbcinst.so /usr/lib/x86_64-linux-gnux32/libxml2.so /usr/lib/x86_64-linux-gnux32/libcrypto.so /usr/lib/x86_64-linux-gnux32/libssl.so /usr/lib/x86_64-linux-gnux32/libzstd.so /usr/lib/x86_64-linux-gnux32/liblzma.so /usr/lib/x86_64-linux-gnux32/libdeflate.so /usr/lib/x86_64-linux-gnux32/liblz4.so /usr/lib/libblosc.so /usr/lib/libarmadillo.so /usr/lib/x86_64-linux-gnux32/libqhull_r.so /usr/lib/x86_64-linux-gnux32/libz.so /usr/lib/x86_64-linux-gnux32/libxerces-c.so /usr/lib/x86_64-linux-gnux32/libjpeg.so /usr/lib/x86_64-linux-gnux32/libz.so /usr/lib/x86_64-linux-gnux32/libtiff.so /usr/lib/x86_64-linux-gnux32/libjpeg.so /usr/lib/x86_64-linux-gnux32/libgeotiff.so /usr/lib/x86_64-linux-gnux32/libtiff.so /usr/lib/x86_64-linux-gnux32/libjpeg.so /usr/lib/x86_64-linux-gnux32/libgeotiff.so /usr/lib/x86_64-linux-gnux32/libtiff.so /usr/lib/x86_64-linux-gnux32/libheif.so /usr/lib/x86_64-linux-gnux32/libjpeg.so /usr/lib/x86_64-linux-gnux32/libpng.so /usr/lib/x86_64-linux-gnux32/libz.so /usr/lib/x86_64-linux-gnux32/libzstd.so /usr/lib/x86_64-linux-gnux32/libpng.so /usr/lib/x86_64-linux-gnux32/libz.so /usr/lib/x86_64-linux-gnux32/libz.so /usr/lib/x86_64-linux-gnux32/libjpeg.so /usr/lib/x86_64-linux-gnux32/libz.so /usr/lib/x86_64-linux-gnux32/libpoppler.so /usr/lib/x86_64-linux-gnux32/libpng.so /usr/lib/x86_64-linux-gnux32/libz.so /usr/lib/x86_64-linux-gnux32/libgif.so /usr/lib/x86_64-linux-gnux32/libnetcdf.so.19 /usr/lib/libblosc.so /usr/lib/x86_64-linux-gnux32/libcfitsio.so /usr/lib/x86_64-linux-gnux32/hdf5/serial/libhdf5.so /usr/lib/x86_64-linux-gnux32/libcrypto.so /usr/lib/x86_64-linux-gnux32/libcurl.so /usr/lib/x86_64-linux-gnux32/libpthread.so /usr/lib/x86_64-linux-gnux32/libsz.so /usr/lib/x86_64-linux-gnux32/libz.so /usr/lib/x86_64-linux-gnux32/libdl.so -lm /usr/lib/x86_64-linux-gnux32/libcurl.so /usr/lib/x86_64-linux-gnux32/libwebp.so /usr/lib/x86_64-linux-gnux32/libsqlite3.so /usr/lib/x86_64-linux-gnux32/libz.so /usr/lib/x86_64-linux-gnux32/libpq.so /usr/lib/x86_64-linux-gnux32/libopenjp2.so /usr/lib/x86_64-linux-gnux32/libxerces-c.so /usr/lib/x86_64-linux-gnux32/libsqlite3.so /usr/lib/x86_64-linux-gnux32/libxerces-c.so /usr/lib/x86_64-linux-gnux32/libkmlbase.so /usr/lib/x86_64-linux-gnux32/libkmldom.so /usr/lib/x86_64-linux-gnux32/libkmlengine.so /usr/lib/x86_64-linux-gnux32/libxerces-c.so /usr/lib/x86_64-linux-gnux32/libfyba.so /usr/lib/x86_64-linux-gnux32/libfygm.so /usr/lib/x86_64-linux-gnux32/libfyut.so /usr/lib/x86_64-linux-gnux32/libsqlite3.so /usr/lib/x86_64-linux-gnux32/libpcre2-8.so /usr/lib/x86_64-linux-gnux32/libspatialite.so /usr/lib/x86_64-linux-gnux32/libsqlite3.so /usr/lib/x86_64-linux-gnux32/libsqlite3.so /usr/lib/x86_64-linux-gnux32/libsqlite3.so /usr/lib/x86_64-linux-gnux32/libsqlite3.so /usr/lib/x86_64-linux-gnux32/libxerces-c.so /usr/lib/x86_64-linux-gnux32/libmysqlclient.so /usr/lib/x86_64-linux-gnux32/libpq.so /usr/lib/x86_64-linux-gnux32/libodbc.so /usr/lib/x86_64-linux-gnux32/libodbc.so /usr/lib/x86_64-linux-gnux32/libodbc.so /usr/lib/x86_64-linux-gnux32/libfreexl.so /usr/lib/libogdi.so /lib/x86_64-linux-gnux32/libexpat.so.1.8.8 /usr/lib/x86_64-linux-gnux32/libxml2.so /usr/lib/x86_64-linux-gnux32/libgeos_c.so.1.16.1 /usr/lib/x86_64-linux-gnux32/libproj.so.25.9.0.0 /usr/lib/x86_64-linux-gnux32/libjson-c.so.5.2.0 -lpthread /usr/lib/x86_64-linux-gnux32/libcrypto.so -lm /usr/lib/libdfalt.so /usr/lib/libmfhdfalt.so 
/usr/bin/ld: gcore/CMakeFiles/gcore.dir/rasterio.cpp.o: in function `void GDALFastCopy<unsigned char>(unsigned char*, int, unsigned char const*, int, int)':
./build-py3.10/gcore/./gcore/rasterio.cpp:3181: undefined reference to `GDALUnrolledCopy_GByte_3_1_SSSE3(unsigned char*, unsigned char const*, int)'
collect2: error: ld returned 1 exit status
make[4]: *** [CMakeFiles/GDAL.dir/build.make:2544: libgdal.so.31.0.0] Error 1
```
x32 architecture is like x86_64 but with 32bits pointers


Full logs:
https://buildd.debian.org/status/fetch.php?pkg=gdal&arch=x32&ver=3.5.0%2Bdfsg-1&stamp=1654937471&raw=0

--- a/gcore/rasterio.cpp
+++ b/gcore/rasterio.cpp
@@ -3117,9 +3117,8 @@ static inline void GDALUnrolledCopy( T* CPL_RESTRICT pDest,
 
 #ifdef HAVE_SSSE3_AT_COMPILE_TIME
 
-void GDALUnrolledCopy_GByte_3_1_SSSE3( GByte* CPL_RESTRICT pDest,
-                                             const GByte* CPL_RESTRICT pSrc,
-                                             GPtrDiff_t nIters );
+#include "rasterio_ssse3.h"
+
 #endif
 
 

--- a/gcore/rasterio_ssse3.cpp
+++ b/gcore/rasterio_ssse3.cpp
@@ -32,16 +32,14 @@ CPL_CVSID("$Id$")
 
 #if defined(HAVE_SSSE3_AT_COMPILE_TIME) && ( defined(__x86_64) || defined(_M_X64) )
 
+#include "rasterio_ssse3.h"
+
 #include <tmmintrin.h>
 #include "gdal_priv_templates.hpp"
 
 void GDALUnrolledCopy_GByte_3_1_SSSE3( GByte* CPL_RESTRICT pDest,
                                              const GByte* CPL_RESTRICT pSrc,
-                                             GInt64 nIters );
-
-void GDALUnrolledCopy_GByte_3_1_SSSE3( GByte* CPL_RESTRICT pDest,
-                                             const GByte* CPL_RESTRICT pSrc,
-                                             GInt64 nIters )
+                                             GPtrDiff_t nIters )
 {
     decltype(nIters) i;
     const __m128i xmm_shuffle0 = _mm_set_epi8(-1  ,-1  ,-1  ,-1,

--- a/gcore/rasterio_ssse3.h
+++ b/gcore/rasterio_ssse3.h
@@ -1,0 +1,42 @@
+/******************************************************************************
+ *
+ * Project:  GDAL Core
+ * Purpose:  SSSE3 specializations
+ * Author:   Even Rouault <even dot rouault at spatialys dot com>
+ *
+ ******************************************************************************
+ * Copyright (c) 2016, Even Rouault <even dot rouault at spatialys dot com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#ifndef RASTERIO_SSSE3_H_INCLUDED
+#define RASTERIO_SSSE3_H_INCLUDED
+
+#include "cpl_port.h"
+
+#if defined(HAVE_SSSE3_AT_COMPILE_TIME) && ( defined(__x86_64) || defined(_M_X64) )
+
+void GDALUnrolledCopy_GByte_3_1_SSSE3( GByte* CPL_RESTRICT pDest,
+                                       const GByte* CPL_RESTRICT pSrc,
+                                       GPtrDiff_t nIters );
+
+#endif
+
+#endif /* RASTERIO_SSSE3_H_INCLUDED */

